specfile/macro_definitions.py

```diff
@@ -0,0 +1,209 @@
+# SPDX-License-Identifier: MIT
+
+import collections
+import re
+from typing import List, Optional, Tuple, Union, overload
+
+from specfile.types import SupportsIndex
+
+
+class MacroDefinition:
+    def __init__(
+        self,
+        name: str,
+        body: str,
+        global_: bool,
+        whitespace: Tuple[str, str, str, str],
+        preceding_lines: Optional[List[str]] = None,
+    ) -> None:
+        self.name = name
+        self.body = body
+        self._global = global_
+        self._whitespace = whitespace
+        self._preceding_lines = (
+            preceding_lines.copy() if preceding_lines is not None else []
+        )
+
+    def __repr__(self) -> str:
+        global_ = repr(self._global)
+        whitespace = repr(self._whitespace)
+        preceding_lines = repr(self._preceding_lines)
+        return (
+            f"MacroDefinition('{self.name}', '{self.body}', {global_}, "
+            f"{whitespace}, {preceding_lines})"
+        )
+
+    def __str__(self) -> str:
+        ws = self._whitespace
+        macro = "%global" if self._global else "%define"
+        return f"{ws[0]}{macro}{ws[1]}{self.name}{ws[2]}{self.body}{ws[3]}"
+
+    def get_raw_data(self) -> List[str]:
+        result = self._preceding_lines.copy()
+        ws = self._whitespace
+        macro = "%global" if self._global else "%define"
+        body = self.body.replace("\n", "\\\n").splitlines()
+        if body:
+            body[-1] += ws[3]
+        else:
+            body = [ws[3]]
+        result.append(f"{ws[0]}{macro}{ws[1]}{self.name}{ws[2]}{body[0]}")
+        result.extend(body[1:])
+        return result
+
+
+class MacroDefinitions(collections.UserList):
+    """
+    Class that represents all macro definitions.
+
+    Attributes:
+        data: List of individual macro definitions.
+    """
+
+    def __init__(
+        self,
+        data: Optional[List[MacroDefinition]] = None,
+        remainder: Optional[List[str]] = None,
+    ) -> None:
+        """
+        Constructs a `MacroDefinitions` object.
+
+        Args:
+            data: List of individual macro definitions.
+            remainder: Leftover lines that can't be parsed into macro definitions.
+
+        Returns:
+            Constructed instance of `MacroDefinitions` class.
+        """
+        super().__init__()
+        if data is not None:
+            self.data = data.copy()
+        self._remainder = remainder.copy() if remainder is not None else []
+
+    def __repr__(self) -> str:
+        data = repr(self.data)
+        remainder = repr(self._remainder)
+        return f"MacroDefinitions({data}, {remainder})"
+
+    def __contains__(self, name: object) -> bool:
+        try:
+            # use parent's __getattribute__() so this method can be called from __getattr__()
+            data = super().__getattribute__("data")
+        except AttributeError:
+            return False
+        return name in [md.name for md in data]
```

Comment:
You don't have to create a whole list before checking, do you?
```suggestion
        return any(md.name == name for md in data)
```